File: python/taflow/metrics/exposure.py
"""Non-zero observation exposure metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import Exposure as _Native
from ._input import as_metric_series


class Exposure:
    """Compute the share of usable periods with non-zero activity.

    ``from_positions`` is the preferred semantic factory: each observation is
    an explicit position, portfolio weight, or exposure-state scalar, exact
    zero means cash/flat, and any finite non-zero value means exposed.
    ``from_returns`` implements QuantStats' weaker proxy contract explicitly:
    a non-zero decimal simple period return is treated as exposed and an exact
    zero as unexposed. It does not claim to reconstruct actual positions; an
    invested asset may genuinely have a zero return.

    The independent oracle is QuantStats 0.0.81 ``exposure`` with return
    preparation disabled. TAFlow deliberately retains its ceiling quirk:
    ``ceil(raw_fraction * 100) / 100`` rounds any fractional result upward to
    the next percentage point. Before that formula, TAFlow applies its package
    missing-value contract, so omitted NaNs are absent from both numerator and
    denominator. During warm-up, an empty or all-omitted stream returns
    ``None`` instead of the oracle's division error. Exact negative zero is
    unexposed; infinities are rejected; return-proxy values below -100% are
    invalid, while explicit positions may be negative or leveraged.

    ``append``, ``extend``, and ``reset`` are fluent. Bulk work releases the
    GIL, while Rust owns the O(1)-memory counters. ``append`` and unchanged
    ``compute`` are O(1), and bulk ``extend`` is one native linear pass.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use a semantic ``from_*`` factory."""
        raise TypeError("use Exposure.from_positions/from_returns")

    @classmethod
    def _create(
        cls,
        values: Any,
        input_mode: str,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "Exposure":
        state = cls.__new__(cls)
        state._state = _Native(input_mode, nan_policy=nan_policy)
        return state.extend(values, column=column)

    @classmethod
    def from_positions(
        cls,
        positions: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "Exposure":
        """Construct from explicit chronological position or exposure states."""
        return cls._create(
            positions, "positions", nan_policy=nan_policy, column=column
        )

    @classmethod
    def from_returns(
        cls,
        returns: Any,
        *,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "Exposure":
        """Construct from the explicit non-zero-return activity proxy."""
        return cls._create(
            returns, "returns", nan_policy=nan_policy, column=column
        )

    def append(self, value: float) -> "Exposure":
        """Append one value in the factory-selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "Exposure":
        """Append a chronological series in the selected domain and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return upward-rounded exposure, or ``None`` without usable periods."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current O(1) scalar without replaying observations."""
        return self._state.compute()

    def reset(self) -> "Exposure":
        """Clear counters, preserve configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable periods counted by Rust."""
        return len(self._state)


__all__ = ["Exposure"]

File: crates/taflow-metrics/src/metrics/exposure.rs
use crate::{MetricError, MetricResult, NanPolicy};

/// Semantic observation domain for [`Exposure`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExposureInputKind {
    /// Period returns, where exact non-zero return is an activity proxy.
    Returns,
    /// Explicit position, weight, or exposure state, where non-zero means invested.
    Positions,
}

/// Share of usable periods with a non-zero return or position state.
#[derive(Debug, Clone)]
pub struct Exposure {
    input_kind: Option<ExposureInputKind>,
    nan_policy: NanPolicy,
    exposed: usize,
    usable: usize,
    position: usize,
}

impl Exposure {
    /// Construct an empty state for an explicit return-proxy or position domain.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input_kind: None,
            nan_policy,
            exposed: 0,
            usable: 0,
            position: 0,
        })
    }

    pub fn from_returns(&mut self, returns: &[f64]) -> MetricResult<&mut Self> {
        self.bind(ExposureInputKind::Returns)?;
        self.extend(returns)?;
        Ok(self)
    }

    pub fn from_positions(&mut self, positions: &[f64]) -> MetricResult<&mut Self> {
        self.bind(ExposureInputKind::Positions)?;
        self.extend(positions)?;
        Ok(self)
    }

    fn bind(&mut self, kind: ExposureInputKind) -> MetricResult<()> {
        match self.input_kind {
            None => self.input_kind = Some(kind),
            Some(selected) if selected == kind => {}
            Some(_) => {
                return Err(MetricError::InvalidParameter {
                    name: "input_kind",
                    value: format!("{kind:?}"),
                    reason: "exposure input domain is already selected",
                })
            }
        }
        Ok(())
    }

    fn ingest(&mut self, value: f64) -> MetricResult<()> {
        let input_kind = self.input_kind.ok_or(MetricError::InvalidParameter {
            name: "input_kind",
            value: "unbound".to_owned(),
            reason: "call from_returns or from_positions before append or extend",
        })?;
        let position = self.position;
        if value.is_nan() {
            if self.nan_policy == NanPolicy::Raise {
                return Err(MetricError::InvalidObservation {
                    domain: match input_kind {
                        ExposureInputKind::Returns => "return",
                        ExposureInputKind::Positions => "position",
                    },
                    position,
                    value: value.to_string(),
                    reason: "NaN is forbidden by nan_policy='raise'",
                });
            }
            self.position += 1;
            return Ok(());
        }
        if !value.is_finite() {
            return Err(MetricError::InvalidObservation {
                domain: match input_kind {
                    ExposureInputKind::Returns => "return",
                    ExposureInputKind::Positions => "position",
                },
                position,
                value: value.to_string(),
                reason: "infinite values are not supported",
            });
        }
        if input_kind == ExposureInputKind::Returns && value < -1.0 {
            return Err(MetricError::InvalidObservation {
                domain: "return",
                position,
                value: value.to_string(),
                reason: "simple returns must be greater than or equal to -1",
            });
        }

        self.exposed += usize::from(value != 0.0);
        self.usable += 1;
        self.position += 1;
        Ok(())
    }

    /// Append one chronological observation and return exposure to date.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        self.ingest(value)?;
        Ok(self.value())
    }

    /// Append a chronological slice through the same bounded native state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            self.ingest(value)?;
        }
        Ok(self.value())
    }

    /// Return exposure rounded upward to the next percentage point.
    ///
    /// The raw fraction is non-zero usable observations divided by all usable
    /// observations. Applying `ceil(fraction * 100) / 100` preserves the
    /// QuantStats 0.0.81 oracle contract. This is O(1) from two counters.
    pub fn value(&self) -> Option<f64> {
        (self.usable != 0).then(|| {
            let raw = self.exposed as f64 / self.usable as f64;
            (raw * 100.0).ceil() / 100.0
        })
    }

    /// Return the current O(1) scalar without replaying observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving domain and missing policy.
    pub fn reset(&mut self) {
        self.exposed = 0;
        self.usable = 0;
        self.position = 0;
    }

    /// Return the number of usable observations processed.
    pub fn len(&self) -> usize {
        self.usable
    }

    /// Return whether no usable observations have been processed.
    pub fn is_empty(&self) -> bool {
        self.usable == 0
    }
}
